app/models/host/base.rb

```diff
@@ -118,17 +137,27 @@ def import_facts(facts)
 
       facts[:domain] = facts[:domain].downcase if facts[:domain].present?
 
+      type = facts.delete(:_type)
+      importer = FactImporter.importer_for(type).new(self, facts)
+      importer.import!
+
+      save(:validate => false)
+
+      parse_facts facts, type, source_proxy
+    end
+
+    def parse_facts(facts, type, source_proxy)
       time = facts[:_timestamp]
       time = time.to_time if time.is_a?(String)
       self.last_compile = time if time
 
-      type = facts.delete(:_type) || 'puppet'
-      importer = FactImporter.importer_for(type).new(self, facts)
-      importer.import!
+      unless build?
+        parser = FactParser.parser_for(type).new(facts)
+
+        populate_fields_from_facts(parser, type, source_proxy)
+      end
 
-      save(:validate => false)
       set_taxonomies(facts)
-      populate_fields_from_facts(facts, type)
```

Comment:
The `set_taxonomies` call needs to be done before `populate_fields_from_facts`. Created https://github.com/theforeman/foreman/pull/6478 to fix this.